File: android/src/main/java/com/neurodoc/TextExtractor.kt
package com.neurodoc

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.pdf.PdfRenderer
import android.os.ParcelFileDescriptor
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import com.tom_roush.pdfbox.pdmodel.PDDocument
import com.tom_roush.pdfbox.text.PDFTextStripper
import com.tom_roush.pdfbox.text.TextPosition
import java.io.File
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine
import kotlin.math.abs

object TextExtractor {

    data class TextBlock(
        val text: String,
        val x: Float,       // normalized 0-1, top-left origin
        val y: Float,
        val width: Float,
        val height: Float,
        val fontSize: Float,
        val fontName: String,
        val confidence: Float
    )

    suspend fun extractText(
        context: Context,
        pdfUrl: String,
        pageIndex: Int,
        mode: String,
        language: String,
        promise: Promise
    ) {
        try {
            val file = resolveFile(pdfUrl)
            val doc = PDDocument.load(file)

            if (pageIndex < 0 || pageIndex >= doc.numberOfPages) {
                doc.close()
                promise.reject("TEXT_EXTRACTION_FAILED", "Invalid page index: $pageIndex")
                return
            }

            val page = doc.getPage(pageIndex)
            val mediaBox = page.mediaBox
            val pageWidth = mediaBox.width
            val pageHeight = mediaBox.height

            val blocks: List<TextBlock>
            val usedMode: String

            when (mode) {
                "native" -> {
                    blocks = extractNativeText(doc, pageIndex, pageWidth, pageHeight)
                    usedMode = "native"
                }
                "ocr" -> {
                    blocks = extractWithOcr(pdfUrl, pageIndex, pageWidth, pageHeight)
                    usedMode = "ocr"
                }
                else -> { // "auto"
                    val nativeBlocks = extractNativeText(doc, pageIndex, pageWidth, pageHeight)
                    if (nativeBlocks.isNotEmpty() && nativeBlocks.any { it.text.trim().isNotEmpty() }) {
                        blocks = nativeBlocks
                        usedMode = "native"
                    } else {
                        blocks = extractWithOcr(pdfUrl, pageIndex, pageWidth, pageHeight)
                        usedMode = "ocr"
                    }
                }
            }

            doc.close()
            promise.resolve(buildResult(blocks, pageWidth, pageHeight, usedMode))
        } catch (e: Exception) {
            promise.reject("TEXT_EXTRACTION_FAILED", e.message, e)
        }
    }

    // MARK: - Native text extraction with PDFBox

    private fun extractNativeText(
        doc: PDDocument,
        pageIndex: Int,
        pageWidth: Float,
        pageHeight: Float
    ): List<TextBlock> {
        val stripper = PositionalTextStripper(pageWidth, pageHeight)
        stripper.startPage = pageIndex + 1 // PDFTextStripper is 1-based
        stripper.endPage = pageIndex + 1
        stripper.getText(doc)

        return groupCharactersIntoWords(stripper.characters, pageWidth, pageHeight)
    }

    private class PositionalTextStripper(
        private val pageWidth: Float,
        private val pageHeight: Float
    ) : PDFTextStripper() {

        val characters = mutableListOf<CharInfo>()

        data class CharInfo(
            val char: String,
            val x: Float,         // absolute PDF coords (bottom-left origin)
            val y: Float,
            val width: Float,
            val height: Float,
            val fontSize: Float,
            val fontName: String
        )

        override fun processTextPosition(text: TextPosition) {
            characters.add(
                CharInfo(
                    char = text.unicode,
                    x = text.xDirAdj,
                    y = text.yDirAdj,
                    width = text.widthDirAdj,
                    height = text.heightDir,
                    fontSize = text.fontSize,
                    fontName = text.font?.name ?: "Unknown"
                )
            )
        }
    }

    private fun groupCharactersIntoWords(
        characters: List<PositionalTextStripper.CharInfo>,
        pageWidth: Float,
        pageHeight: Float
    ): List<TextBlock> {
        if (characters.isEmpty()) return emptyList()

        val words = mutableListOf<TextBlock>()
        var currentText = StringBuilder()
        var currentX = characters[0].x
        var currentY = characters[0].y
        var currentMaxX = characters[0].x + characters[0].width
        var currentHeight = characters[0].height
        var currentFontSize = characters[0].fontSize
        var currentFontName = characters[0].fontName

        for (i in characters.indices) {
            val char = characters[i]
            val isWhitespace = char.char.isBlank()

            if (i > 0) {
                val prev = characters[i - 1]
                val isNewLine = abs(char.y - prev.y) > prev.height * 0.5f
                val gap = char.x - (prev.x + prev.width)
                val isLargeGap = gap > prev.width * 0.4f

                if (isWhitespace || isNewLine || isLargeGap) {
                    // Save current word
                    val text = currentText.toString().trim()
                    if (text.isNotEmpty()) {
                        words.add(toNormalizedBlock(
                            text, currentX, currentY, currentMaxX - currentX, currentHeight,
                            currentFontSize, currentFontName, pageWidth, pageHeight
                        ))
                    }

                    if (!isWhitespace) {
                        currentText = StringBuilder(char.char)
                        currentX = char.x
                        currentY = char.y
                        currentMaxX = char.x + char.width
                        currentHeight = char.height
                        currentFontSize = char.fontSize
                        currentFontName = char.fontName
                    } else {
                        currentText = StringBuilder()
                        currentX = 0f
                        currentY = 0f
                        currentMaxX = 0f
                        currentHeight = 0f
                    }
                    continue
                }
            }

            if (!isWhitespace) {
                if (currentText.isEmpty()) {
                    currentX = char.x
                    currentY = char.y
                    currentMaxX = char.x + char.width
                    currentHeight = char.height
                    currentFontSize = char.fontSize
                    currentFontName = char.fontName
                } else {
                    currentMaxX = char.x + char.width
                    if (char.height > currentHeight) currentHeight = char.height
                }
                currentText.append(char.char)
            }
        }

        // Last word
        val text = currentText.toString().trim()
        if (text.isNotEmpty()) {
            words.add(toNormalizedBlock(
                text, currentX, currentY, currentMaxX - currentX, currentHeight,
                currentFontSize, currentFontName, pageWidth, pageHeight
            ))
        }

        return words
    }

    /**
     * Convert PDFTextStripper coordinates to normalized 0-1, top-left origin.
     * PDFTextStripper's yDirAdj is already in top-left coordinate system (distance from top).
     */
    private fun toNormalizedBlock(
        text: String,
        x: Float,
        y: Float,
        width: Float,
        height: Float,
        fontSize: Float,
        fontName: String,
        pageWidth: Float,
        pageHeight: Float
    ): TextBlock {
        val nx = x / pageWidth
        // yDirAdj is from top, but it's the baseline; adjust to top of text
        val ny = (y - height) / pageHeight
        val nw = width / pageWidth
        val nh = height / pageHeight

        return TextBlock(
            text = text,
            x = nx.coerceIn(0f, 1f),
            y = ny.coerceIn(0f, 1f),
            width = nw.coerceIn(0f, 1f),
            height = nh.coerceIn(0f, 1f),
            fontSize = fontSize,
            fontName = fontName,
            confidence = 1.0f
        )
    }

    // MARK: - OCR fallback

    private suspend fun extractWithOcr(
        pdfUrl: String,
        pageIndex: Int,
        pageWidth: Float,
        pageHeight: Float
    ): List<TextBlock> {
        val bitmap = renderPageToBitmap(pdfUrl, pageIndex)
        val blocks = performOcr(bitmap, pageWidth, pageHeight)
        bitmap.recycle()
        return blocks
    }

    private suspend fun performOcr(
        bitmap: Bitmap,
        pageWidth: Float,
        pageHeight: Float
    ): List<TextBlock> = suspendCoroutine { cont ->
        val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
        val inputImage = InputImage.fromBitmap(bitmap, 0)

        recognizer.process(inputImage)
            .addOnSuccessListener { text ->
                val blocks = mutableListOf<TextBlock>()
                val imgW = bitmap.width.toFloat()
                val imgH = bitmap.height.toFloat()

                for (block in text.textBlocks) {
                    val box = block.boundingBox ?: continue
                    val nx = box.left / imgW
                    val ny = box.top / imgH
                    val nw = box.width() / imgW
                    val nh = box.height() / imgH

                    val fontSize = nh * pageHeight * 0.85f

                    blocks.add(TextBlock(
                        text = block.text,
                        x = nx,
                        y = ny,
                        width = nw,
                        height = nh,
                        fontSize = fontSize,
                        fontName = "Unknown",
                        confidence = block.lines.firstOrNull()?.confidence ?: 0f
                    ))
                }

                cont.resume(blocks)
            }
            .addOnFailureListener { e ->
                cont.resumeWithException(e)
            }
    }

    // MARK: - Helpers

    private fun renderPageToBitmap(pdfUrl: String, pageIndex: Int, scale: Float = 2f): Bitmap {
        val path = if (pdfUrl.startsWith("file://")) pdfUrl.removePrefix("file://") else pdfUrl
        val file = File(path)
        val fd = ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
        val renderer = PdfRenderer(fd)

        val page = renderer.openPage(pageIndex)
        val width = (page.width * scale).toInt()
        val height = (page.height * scale).toInt()

        val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        canvas.drawColor(Color.WHITE)
        page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)

        page.close()
        renderer.close()
        fd.close()

        return bitmap
    }

    private fun resolveFile(urlString: String): File {
        val path = if (urlString.startsWith("file://")) urlString.removePrefix("file://") else urlString
        return File(path)
    }

    private fun buildResult(
        blocks: List<TextBlock>,
        pageWidth: Float,
        pageHeight: Float,
        mode: String
    ): WritableNativeMap {
        val blocksArray = WritableNativeArray()
        for (block in blocks) {
            blocksArray.pushMap(WritableNativeMap().apply {
                putString("text", block.text)
                putMap("boundingBox", WritableNativeMap().apply {
                    putDouble("x", block.x.toDouble())
                    putDouble("y", block.y.toDouble())
                    putDouble("width", block.width.toDouble())
                    putDouble("height", block.height.toDouble())
                })
                putDouble("fontSize", block.fontSize.toDouble())
                putString("fontName", block.fontName)
                putDouble("confidence", block.confidence.toDouble())
            })
        }

        return WritableNativeMap().apply {
            putArray("textBlocks", blocksArray)
            putDouble("pageWidth", pageWidth.toDouble())
            putDouble("pageHeight", pageHeight.toDouble())
            putString("mode", mode)
        }
    }
}

File: ios/TextExtractor.swift
import Foundation
import PDFKit
import Vision
import UIKit

class TextExtractor {

    struct TextBlock {
        let text: String
        let x: Double      // normalized 0-1, top-left origin
        let y: Double
        let width: Double
        let height: Double
        let fontSize: Double
        let fontName: String
        let confidence: Double
    }

    static func extractText(
        pdfUrl: String,
        pageIndex: Int,
        mode: String,
        language: String,
        resolver: @escaping RNResolver,
        rejecter: @escaping RNRejecter
    ) {
        guard let url = resolveUrl(pdfUrl), let doc = PDFDocument(url: url) else {
            rejecter("TEXT_EXTRACTION_FAILED", "Failed to load PDF", nil)
            return
        }

        guard pageIndex >= 0 && pageIndex < doc.pageCount, let page = doc.page(at: pageIndex) else {
            rejecter("TEXT_EXTRACTION_FAILED", "Invalid page index: \(pageIndex)", nil)
            return
        }

        let bounds = page.bounds(for: .mediaBox)
        let pageWidth = Double(bounds.width)
        let pageHeight = Double(bounds.height)

        switch mode {
        case "native":
            let blocks = extractNativeText(page: page, pageWidth: pageWidth, pageHeight: pageHeight)
            resolver(buildResult(blocks: blocks, pageWidth: pageWidth, pageHeight: pageHeight, mode: "native"))

        case "ocr":
            let blocks = extractWithOcr(page: page, pageWidth: pageWidth, pageHeight: pageHeight, language: language)
            resolver(buildResult(blocks: blocks, pageWidth: pageWidth, pageHeight: pageHeight, mode: "ocr"))

        default: // "auto"
            let nativeBlocks = extractNativeText(page: page, pageWidth: pageWidth, pageHeight: pageHeight)
            let hasText = nativeBlocks.contains { !$0.text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

            if hasText && nativeBlocks.count > 0 {
                resolver(buildResult(blocks: nativeBlocks, pageWidth: pageWidth, pageHeight: pageHeight, mode: "native"))
            } else {
                let ocrBlocks = extractWithOcr(page: page, pageWidth: pageWidth, pageHeight: pageHeight, language: language)
                resolver(buildResult(blocks: ocrBlocks, pageWidth: pageWidth, pageHeight: pageHeight, mode: "ocr"))
            }
        }
    }

    // MARK: - Native text extraction

    private static func extractNativeText(page: PDFPage, pageWidth: Double, pageHeight: Double) -> [TextBlock] {
        guard let pageString = page.string, !pageString.isEmpty else {
            return []
        }

        // Get character-level bounds and group into words
        var characters: [(char: Character, bounds: CGRect)] = []

        for i in 0..<pageString.count {
            let selection = page.selection(for: NSRange(location: i, length: 1))
            guard let sel = selection else { continue }

            let selBounds = sel.bounds(for: page)
            guard selBounds.width > 0 && selBounds.height > 0 else { continue }

            let charIndex = pageString.index(pageString.startIndex, offsetBy: i)
            characters.append((char: pageString[charIndex], bounds: selBounds))
        }

        return groupCharactersIntoWords(characters: characters, pageWidth: pageWidth, pageHeight: pageHeight)
    }

    private static func groupCharactersIntoWords(
        characters: [(char: Character, bounds: CGRect)],
        pageWidth: Double,
        pageHeight: Double
    ) -> [TextBlock] {
        guard !characters.isEmpty else { return [] }

        var words: [TextBlock] = []
        var currentWord = String(characters[0].char)
        var currentBounds = characters[0].bounds

        for i in 1..<characters.count {
            let char = characters[i]
            let prev = characters[i - 1]

            // Start new word on whitespace, large horizontal gap, or line change
            let isWhitespace = char.char.isWhitespace || char.char.isNewline
            let isNewLine = abs(char.bounds.midY - prev.bounds.midY) > prev.bounds.height * 0.5
            let gap = char.bounds.minX - prev.bounds.maxX
            let isLargeGap = gap > prev.bounds.width * 0.4

            if isWhitespace || isNewLine || isLargeGap {
                // Save current word if non-empty
                let trimmed = currentWord.trimmingCharacters(in: .whitespaces)
                if !trimmed.isEmpty {
                    words.append(boundsToBlock(text: trimmed, bounds: currentBounds, pageWidth: pageWidth, pageHeight: pageHeight))
                }

                if !isWhitespace {
                    // Start new word with this character
                    currentWord = String(char.char)
                    currentBounds = char.bounds
                } else {
                    // Reset for next non-whitespace character
                    currentWord = ""
                    currentBounds = .zero
                }
            } else {
                currentWord.append(char.char)
                currentBounds = currentBounds.union(char.bounds)
            }
        }

        // Don't forget the last word
        let trimmed = currentWord.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty {
            words.append(boundsToBlock(text: trimmed, bounds: currentBounds, pageWidth: pageWidth, pageHeight: pageHeight))
        }

        return words
    }

    /// Convert PDFKit bounds (bottom-left origin, points) to normalized block (top-left origin, 0-1)
    private static func boundsToBlock(text: String, bounds: CGRect, pageWidth: Double, pageHeight: Double) -> TextBlock {
        let nx = Double(bounds.origin.x) / pageWidth
        // PDFKit: origin.y is bottom edge, we need top edge
        let ny = 1.0 - (Double(bounds.origin.y) + Double(bounds.height)) / pageHeight
        let nw = Double(bounds.width) / pageWidth
        let nh = Double(bounds.height) / pageHeight

        // Estimate font size from character height (~80% of bounding box height)
        let fontSize = Double(bounds.height) * 0.85

        return TextBlock(
            text: text,
            x: nx,
            y: ny,
            width: nw,
            height: nh,
            fontSize: fontSize,
            fontName: "Unknown",
            confidence: 1.0
        )
    }

    // MARK: - OCR fallback

    private static func extractWithOcr(page: PDFPage, pageWidth: Double, pageHeight: Double, language: String) -> [TextBlock] {
        guard let cgImage = renderPageToImage(page: page) else { return [] }

        var blocks: [TextBlock] = []

        let request = VNRecognizeTextRequest { request, _ in
            if let observations = request.results as? [VNRecognizedTextObservation] {
                for observation in observations {
                    guard let candidate = observation.topCandidates(1).first else { continue }

                    let box = observation.boundingBox
                    // Vision: bottom-left origin, normalized 0-1
                    let nx = box.origin.x
                    let ny = 1.0 - box.origin.y - box.height
                    let nw = box.width
                    let nh = box.height

                    let fontSize = Double(nh) * pageHeight * 0.85

                    blocks.append(TextBlock(
                        text: candidate.string,
                        x: nx,
                        y: ny,
                        width: nw,
                        height: nh,
                        fontSize: fontSize,
                        fontName: "Unknown",
                        confidence: Double(candidate.confidence)
                    ))
                }
            }
        }

        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true
        if language != "auto" {
            request.recognitionLanguages = [language]
        }

        // VNImageRequestHandler.perform is synchronous — completion is called before perform returns
        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        try? handler.perform([request])

        return blocks
    }

    // MARK: - Helpers

    private static func buildResult(blocks: [TextBlock], pageWidth: Double, pageHeight: Double, mode: String) -> [String: Any] {
        let blocksArray: [[String: Any]] = blocks.map { block in
            [
                "text": block.text,
                "boundingBox": [
                    "x": block.x,
                    "y": block.y,
                    "width": block.width,
                    "height": block.height,
                ] as [String: Any],
                "fontSize": block.fontSize,
                "fontName": block.fontName,
                "confidence": block.confidence,
            ] as [String: Any]
        }

        return [
            "textBlocks": blocksArray,
            "pageWidth": pageWidth,
            "pageHeight": pageHeight,
            "mode": mode,
        ] as [String: Any]
    }
}
